src/ir/ir_traversing_visitor_with_result.h

```diff
@@ -0,0 +1,80 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#ifndef SRC_IR_IR_TRAVERSING_VISITOR_WITH_RESULT_H_
+#define SRC_IR_IR_TRAVERSING_VISITOR_WITH_RESULT_H_
+
+#include <variant>
+#include "src/ir/ir_visitor.h"
+#include "src/ir/module.h"
+
+namespace raksha::ir {
+
+// A visitor that also traverses the children of a node and allows performing
+// different actions before (PreVisit) and after (PostVisit) the children are
+// visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
+template <typename Derived, typename Result=std::monostate>
+class IRTraversingVisitorWithResult : public IRVisitor<Derived, Result> {
+ public:
+  virtual ~IRTraversingVisitorWithResult() {}
+
+  // Invoked before all the children of `module` is visited.
+  virtual Result PreVisit(const Module& module) { return Result(); }
+  // Invoked after all the children of `module` is visited.
+  virtual void PostVisit(const Module& module, Result& result) {}
+  virtual void FoldResult(const Module& module, Result child_result, Result& result) {}
```

Comment:
Don't we need the parent `Operation` as argument as well?

Reply:
Similar to above.

Reply:
Will continue discussion in the other thread.